HDDS-12810. Check and reserve space atomically in VolumeChoosingPolicy

peterxcli opened the pull request:
## What changes were proposed in this pull request?

If `VolumeChoosingPolicy#chooseVolume` is not synchronized and caller required space reserved before chooseVolume returns, there is no easy way to guarantee chosen volume will have enough space.

`VolumeChoosingPolicy#chooseVolume` is used for create container and container import. Make `VolumeChoosingPolicy#chooseVolume` synchronized or part of chooseVolume synchronized will add some latency, which is negligible for container import, may noticeable for create container. We can compare the create container latency metrics with and without the synchronization.

`VolumeChoosingPolicy#chooseVolume`, its Java Doc says "The implementations of this interface must be thread-safe.", regarding the space full check and space reservation, it can be done as an atomic operation in the chooseVolume internally, so that there will no over allocation of space due to concurrent container creation and container import.

Also need to add test to check if committedBytes is checked in container creation unit tests and container import, container replication unit tests. Besides, we also need the unit tests to verify that if the container creation fails, container import fails, or container replication fails, the resevered committedBytes are released.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-12810

## How was this patch tested?

Added coverage of commit space would be reserved/released correctly when creating/importing containers in `TestKeyValueContainer`, `TestDownloadAndImportReplicator` and `TestSendContainerRequestHandler`. There're also some existing tests covering this change, like `TestContainerPerisistance` and `TestOzoneContainer`.

CI:
~~https://github.com/peterxcli/ozone/actions/runs/14911933939~~
https://github.com/peterxcli/ozone/actions/runs/15001277972

peterxcli committed:
Make volume space check and reservation as an atomic operation in VolumeChoosingPolicy#chooseVolume

peterxcli committed:
Explictly reserve/release commit space

peterxcli committed:
Add coverge of commit space reserve/release cases for container importing

peterxcli committed:
Add tests for container creation commit space reservation and exception handling

peterxcli commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/SendContainerRequestHandler.java:
```diff
@@ -87,16 +86,6 @@ public void onNext(SendContainerRequest req) {
-        volume.incCommittedBytes(importer.getDefaultContainerSize() * 2);
-        StorageLocationReport volumeReport = volume.getReport();
-        // Already committed bytes increased above, so required space is not required here in AvailableSpaceFilter
-        if (volumeReport.getUsableSpace() <= 0) {
```
This step(increase commit space the check usable space) has been done in the `VolumeChoosingPolicy#chooseVolume`, so removed

peterxcli commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/DownloadAndImportReplicator.java:
```diff
@@ -81,15 +81,6 @@ public void replicate(ReplicationTask task) {
     try {
       targetVolume = containerImporter.chooseNextVolume();
-      // Increment committed bytes and verify if it doesn't cross the space left.
-      targetVolume.incCommittedBytes(containerSize * 2);
```
This step(increase commit space the check usable space) has been done in the VolumeChoosingPolicy#chooseVolume, so removed, too

peterxcli committed:
original behaviour didnt consider container state when committing space

peterxcli commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerReader.java:
```diff
@@ -202,50 +202,52 @@ public void verifyAndFixupContainerData(ContainerData containerData)
+      }
+
+      try {
+        containerData.commitSpace();
```
Sorry this diff may hard to read, but I'm doing here is just:
1. Move the `containerData` instance check up to reduce the indent level of the block.
2. Commit space before adding container. **But I have one question here, IIRC, only containers in RECOVERING or OPEN will receive new write, so would it be better to only commit the space for RECOVERING/OPEN containers?**

siddhantsangwan replied:
What's the logic behind writing to `RECOVERING` containers? AFAIK they shouldn't be written to. I think the current code in the PR is fine since the `commitSpace()` method already checks whether the container's state is `OPEN`. Maybe we can add a comment before line 239 that says this.

peterxcli commented:
I noticed that the push/pull replicator reserves twice the container size retrieved from the local configuration (`OZONE_SCM_CONTAINER_SIZE`). I was wondering if it might be more appropriate to use the maximum size of the replicated container instead, since the container size config may vary across different nodes.

@ChenSammi, what are your thoughts on this approach? Does it seem reasonable to you?

siddhantsangwan commented:
> I was wondering if it might be more appropriate to use the maximum size of the replicated container instead

Good point to use the actual container size instead, but we don't know what the actual size is (when committing space) before actually downloading the container, right?

siddhantsangwan commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerReader.java:
```diff
@@ -202,50 +202,52 @@ public void verifyAndFixupContainerData(ContainerData containerData)
+        containerData.commitSpace();
+        containerSet.addContainer(kvContainer);
+      } catch (StorageContainerException e) {
+        if (e.getResult() != ContainerProtos.Result.CONTAINER_EXISTS) {
+          throw e;
+        }
```
I think we need to release committed space in this case as well.

siddhantsangwan replied:
It'd be good to have testing for the new changes in `ContainerReader` as well.

siddhantsangwan commented:
@peterxcli Thanks for taking this up, it's a good fix. I've left a few review comments.

ChenSammi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/SendContainerRequestHandler.java:
```diff
@@ -130,7 +119,7 @@ public void onError(Throwable t) {
     } finally {
       if (volume != null) {
-        volume.incCommittedBytes(-importer.getDefaultContainerSize() * 2);
+        volume.incCommittedBytes(-HddsServerUtil.requiredReplicationSpace(importer.getDefaultContainerSize()));
```
Can we add one function in ContainerImporter which will return the value of  HddsServerUtil.requiredReplicationSpace(importer.getDefaultContainerSize()), and call this function where it's needed?

ChenSammi commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/replication/TestSendContainerRequestHandler.java:
```diff
@@ -99,4 +115,86 @@ void testReceiveDataForExistingContainer() throws Exception {
+    assertEquals(volume.getCommittedBytes(), initialCommittedBytes + 2 * containerMaxSize);
+
+    // Send a failed response with wrong offset
+    request = createRequest(containerId, ByteString.copyFromUtf8("test"), 0);
```
This request looks the same as previous request.  How to tell it will have a failed response?

ChenSammi replied:
Oh, I see. Could you use the same way as testSpaceReservedAndReleasedWhenOnCompletedFails to mock some failure instead?  This current logic is confusing.

ChenSammi commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/replication/TestSendContainerRequestHandler.java:
```diff
@@ -99,4 +115,86 @@ void testReceiveDataForExistingContainer() throws Exception {
+    sendContainerRequestHandler.onNext(request);
+
+    // Verify commit space is released
+    assertEquals(volume.getCommittedBytes(), initialCommittedBytes);
```
This doesn't seem right.  Would it be initialCommittedBytes + 2 * containerMaxSize?

peterxcli committed:
set commit right after the volume is choosen and handle release case at handler

peterxcli committed:
container reader commit space after adding container and some refactor

peterxcli committed:
Add test in TestContainerReader

peterxcli commented:
> Good point to use the actual container size instead, but we don't know what the actual size is (when committing space) before actually downloading the container, right?

Right, so I think that info should be brought in proto message.

peterxcli commented:
@siddhantsangwan @ChenSammi I have addressed all review comments, please take another look. Thanks!

ChenSammi commented:
> > Good point to use the actual container size instead, but we don't know what the actual size is (when committing space) before actually downloading the container, right?
> 
> Right, so I think that info should be brought in proto message.


Agree. We can do this in a new JIRA.  Generally, the tar zip file of container will have smaller size than it's real size, so reserve a 2* max container size is conservative enough in most cases. But there are container over-allocated case, where container size can be double or triple the max container size as @siddhantsangwan saw in user's environment, have a accurate container size can handle this case welly.

ChenSammi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:
```diff
@@ -434,8 +435,12 @@ ContainerCommandResponseProto handleCreateContainer(
+      exceptionThrown = true;
       return ContainerUtils.logAndReturnError(LOG, ex, request);
     } finally {
+      if (exceptionThrown) {
```
1.  please move this newContainerData.releaseCommitSpace() into the exception handling directly. 
2. we might first check whether space is committed before release, for newContainer.create can throw out exception from volumeChoosingPolicy.chooseVolume  too.

ChenSammi replied:
Okay.

ChenSammi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerReader.java:
```diff
@@ -202,50 +202,53 @@ public void verifyAndFixupContainerData(ContainerData containerData)
+
+      try {
+        containerSet.addContainer(kvContainer);
+
```
This blank line ca be removed.

ChenSammi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/DownloadAndImportReplicator.java:
```diff
@@ -114,7 +105,7 @@ public void replicate(ReplicationTask task) {
     } finally {
       if (targetVolume != null) {
-        targetVolume.incCommittedBytes(-containerSize * 2);
+        targetVolume.incCommittedBytes(-HddsServerUtil.requiredReplicationSpace(containerSize));
```
The change is missed here.

ChenSammi commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/ozoneimpl/TestContainerReader.java:
```diff
@@ -229,21 +239,23 @@ public void testContainerReader(ContainerTestVersionInfo versionInfo)
       //recovering container should be marked unhealthy, so the count should be 3
-      assertEquals(UNHEALTHY, containerSet.getContainer(
-          recoveringContainerData.getContainerID()).getContainerState());
+      Container c = containerSet.getContainer(
```
Looks like the recoveringContainerData replicationIndex is not explicitly set, which means it's 0. 

So this else branch doesn't have the chance to get executed.  Besides, L258 conflicts with L278.

peterxcli committed:
Addressed comments

peterxcli committed:
Merge remote-tracking branch 'upstream/master' into hdds12810-make-volume-space-check-and-reservation-as-an-atomic-operation-in-volume-choosing-policy

peterxcli committed:
Move release commit byte check logic to TestKeyValueHandler

ChenSammi commented:
@peterxcli , can we only measure the time spend by 

```
for (int j = 0; j < NUM_ITERATIONS; j++) {
            try {
              HddsVolume volume = policy.chooseVolume(volumes, CONTAINER_SIZE);
              if (volume != null) {
                successCount.incrementAndGet();
              }
            } catch (Exception e) {
              failureCount.incrementAndGet();
            }
          }
```

So we can avoid to count the thread start/stop time?  The two polices, they are expected to have the same trend.

peterxcli commented:
Updated result:
<img width="991" alt="image" src="https://github.com/user-attachments/assets/21b6eb72-f292-4fee-a275-c36c05d9e461" />


<details>
  <summary>test script</summary>

```java
public class TestVolumeChoosingPolicyPerformance {
  private static final Logger LOG = LoggerFactory.getLogger(TestVolumeChoosingPolicyPerformance.class);
  private static final OzoneConfiguration CONF = new OzoneConfiguration();
  private static final int NUM_VOLUMES = 10;
  private static final int NUM_THREADS = 100;
  private static final int NUM_ITERATIONS = 1000;
  private static final long CONTAINER_SIZE = 1;

  @TempDir
  private Path baseDir;

  private List<HddsVolume> volumes;
  private CapacityVolumeChoosingPolicy capacityPolicy;
  private RoundRobinVolumeChoosingPolicy roundRobinPolicy;

  // @BeforeEach
  public void setup() throws Exception {
    volumes = new ArrayList<>();
    capacityPolicy = new CapacityVolumeChoosingPolicy();
    roundRobinPolicy = new RoundRobinVolumeChoosingPolicy();

    // Create NUM_VOLUMES volumes with different available space
    for (int i = 0; i < NUM_VOLUMES; i++) {
      String volumePath = baseDir + "/disk" + i;
      // Each volume has 1000 bytes capacity and different available space
      SpaceUsageSource source = MockSpaceUsageSource.fixed(1000000000, 1000000000 - i * 50);
      SpaceUsageCheckFactory factory = MockSpaceUsageCheckFactory.of(
          source, Duration.ZERO, SpaceUsagePersistence.None.INSTANCE);
      HddsVolume volume = new HddsVolume.Builder(volumePath)
          .conf(CONF)
          .usageCheckFactory(factory)
          .build();
      volumes.add(volume);
    }
  }

  // @AfterEach
  public void cleanUp() {
    volumes.forEach(HddsVolume::shutdown);
  }

  @Test
  public void testConcurrentVolumeChoosing() throws Exception {
    // Test both policies
    for (int i = 0; i < 5; i++) {
      setup();
      testPolicyPerformance("CapacityVolumeChoosingPolicy", capacityPolicy);
      cleanUp();

      setup();
      testPolicyPerformance("RoundRobinVolumeChoosingPolicy", roundRobinPolicy);
      cleanUp();
    }
  }

  private void testPolicyPerformance(String policyName, VolumeChoosingPolicy policy) throws Exception {
    ExecutorService executor = Executors.newFixedThreadPool(NUM_THREADS);
    CountDownLatch latch = new CountDownLatch(NUM_THREADS);
    AtomicInteger successCount = new AtomicInteger(0);
    AtomicInteger failureCount = new AtomicInteger(0);
    AtomicLong totalTimeNanos = new AtomicLong(0);

    // Submit NUM_THREADS tasks, each doing NUM_ITERATIONS volume choices
    for (int i = 0; i < NUM_THREADS; i++) {
      executor.submit(() -> {
        try {
          long threadStartTime = System.nanoTime();
          for (int j = 0; j < NUM_ITERATIONS; j++) {
            try {
              HddsVolume volume = policy.chooseVolume(volumes, CONTAINER_SIZE);
              if (volume != null) {
                successCount.incrementAndGet();
              }
            } catch (Exception e) {
              failureCount.incrementAndGet();
            }
          }
          long threadEndTime = System.nanoTime();
          totalTimeNanos.addAndGet((threadEndTime - threadStartTime));
        } finally {
          latch.countDown();
        }
      });
    }

    // Wait for all tasks to complete
    assertTrue(latch.await(5, TimeUnit.MINUTES), "Test timed out");
    executor.shutdown();
    assertTrue(executor.awaitTermination(1, TimeUnit.MINUTES), "Executor did not terminate");

    long totalOperations = (long) NUM_THREADS * NUM_ITERATIONS;
    double avgTimePerOperation = (double) totalTimeNanos.get() / totalOperations;
    double operationsPerSecond = totalOperations / (totalTimeNanos.get() / 1_000_000_000.0);

    LOG.info("Performance results for {}:", policyName);
    LOG.info("Total operations: {}", totalOperations);
    LOG.info("Successful operations: {}", successCount.get());
    LOG.info("Failed operations: {}", failureCount.get());
    LOG.info("Total time: {} ms", totalTimeNanos.get() / 1_000_000);
    LOG.info("Average time per operation: {} ns", avgTimePerOperation);
    LOG.info("Operations per second: {}", operationsPerSecond);
  }
}

```

</details>

ChenSammi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java:
```diff
@@ -214,21 +213,14 @@ public synchronized void setState(ContainerDataProto.State state) {
-  void setCommittedSpace(boolean committedSpace) {
-    this.committedSpace = committedSpace;
+  public void setCommittedSpace(boolean committed) {
+    committedSpace = committed;
```
@peterxcli , I remember previous version has committed check to avoid commit twice.  Is it removed?

peterxcli replied:
Yes, otherwise it would lead the test failed, and other reason to remove it is I think it's not harmful to set the same value.
The `ContainerData#commitSpace` has that check because it maybe increase the volume commit bytes more than once.

ChenSammi replied:
Okay.

ChenSammi commented:
The second round data, why the roundrobin policy with synchronized acts better than unsynchronized, do you have any idea? @peterxcli

ChenSammi commented:
In this second run, the capacity policy, it has better performance with synchronized than without synchronized, kind of unexpected.

siddhantsangwan commented:
> In this second run, the capacity policy, it has better performance with synchronized than without synchronized, kind of unexpected.

Yeah, interesting result.

Aside from this, is 100 threads too much? Maybe we should try with 10 threads to see if too much context switching is happening with 100 threads.

peterxcli commented:
> In this second run, the capacity policy, it has better performance with synchronized than without synchronized, kind of unexpected.

My initial guess was that this result was caused by the randomness of the object, but it turns out it may have just been affected by an outlier.![image](https://github.com/user-attachments/assets/226fb536-1222-451c-8edf-f153114ab063)

ChenSammi commented:
I used your code and increase the iteration from 1000 to 1000000.  When iteration is smaller, the variance is huge.  With 1000000 iteration, the data is more consistent.  Here are five run data. 

```
Without patch 

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8595214 ms
Average time per operation: 85952.14961086 ns
Operations per second: 11634.380344498686

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1842511 ms
Average time per operation: 18425.11104205 ns
Operations per second: 54273.75703287695

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8597296 ms
Average time per operation: 85972.96259339 ns
Operations per second: 11631.563806048072

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1705104 ms
Average time per operation: 17051.04959121 ns
Operations per second: 58647.41608138368

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8606974 ms
Average time per operation: 86069.74989122 ns
Operations per second: 11618.483860634644

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1619432 ms
Average time per operation: 16194.32189543 ns
Operations per second: 61750.038467630904

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8241240 ms
Average time per operation: 82412.40437041 ns
Operations per second: 12134.095681827332

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1678796 ms
Average time per operation: 16787.96074832 ns
Operations per second: 59566.49619281912

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8037039 ms
Average time per operation: 80370.39892128 ns
Operations per second: 12442.3918933072

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1641007 ms
Average time per operation: 16410.07820917 ns
Operations per second: 60938.16173534122



With the patch

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8595214 ms
Average time per operation: 85952.14961086 ns
Operations per second: 11634.380344498686

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1842511 ms
Average time per operation: 18425.11104205 ns
Operations per second: 54273.75703287695

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8597296 ms
Average time per operation: 85972.96259339 ns
Operations per second: 11631.563806048072

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1705104 ms
Average time per operation: 17051.04959121 ns
Operations per second: 58647.41608138368

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8606974 ms
Average time per operation: 86069.74989122 ns
Operations per second: 11618.483860634644

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1619432 ms
Average time per operation: 16194.32189543 ns
Operations per second: 61750.038467630904

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8241240 ms
Average time per operation: 82412.40437041 ns
Operations per second: 12134.095681827332

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1678796 ms
Average time per operation: 16787.96074832 ns
Operations per second: 59566.49619281912

Performance results for CapacityVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 8037039 ms
Average time per operation: 80370.39892128 ns
Operations per second: 12442.3918933072

Performance results for RoundRobinVolumeChoosingPolicy
Total operations: 100000000
Successful operations: 100000000
Failed operations: 0
Total time: 1641007 ms
Average time per operation: 16410.07820917 ns
Operations per second: 60938.16173534122

```

peterxcli commented:
@ChenSammi Looks like you pasted the same result.

ChenSammi commented:
I updated the data. Interesting thing is this time roundrobin-sync is better than roundrobin. 

![Screenshot 2025-05-08 at 16 59 04](https://github.com/user-attachments/assets/2b498906-b47d-4e31-a25f-d5bddb0b2ae2)

ChenSammi commented:
> My initial guess was that this result was caused by the random object, but it turns out it may have just been affected by some outliers.

Yes, replace Random with ThreadLocalRandom have greatly improve the performance in my test. After replace using ThreadLocalRandom, we can achieve almost same performance as without synchronized for CapacityVolumeChoosingPolicy.  Could you implement this optimization with a new commit? 


This is the version I used, based on your code, where a single ExecutorService is used for the entire test, and local int is used to record success and failure count in each thread, before aggregate to the global counter. 

```
package org.apache.hadoop.ozone.container.common.volume;

import org.apache.hadoop.hdds.conf.OzoneConfiguration;
import org.apache.hadoop.hdds.fs.MockSpaceUsageCheckFactory;
import org.apache.hadoop.hdds.fs.MockSpaceUsageSource;
import org.apache.hadoop.hdds.fs.SpaceUsageCheckFactory;
import org.apache.hadoop.hdds.fs.SpaceUsagePersistence;
import org.apache.hadoop.hdds.fs.SpaceUsageSource;
import org.apache.hadoop.ozone.container.common.interfaces.VolumeChoosingPolicy;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class TestVolumeChoosingPolicyPerformance {
  private static final OzoneConfiguration CONF = new OzoneConfiguration();
  private static final int NUM_VOLUMES = 10;
  private static final int NUM_THREADS = 100;
  private static final int NUM_ITERATIONS = 1000000;
  private static final long CONTAINER_SIZE = 1;

  @TempDir
  private Path baseDir;

  private List<HddsVolume> volumes;
  private CapacityVolumeChoosingPolicy capacityPolicy;
  private RoundRobinVolumeChoosingPolicy roundRobinPolicy;
  ExecutorService executor = Executors.newFixedThreadPool(NUM_THREADS);

  // @BeforeEach
  public void setup() throws Exception {
    volumes = new ArrayList<>();
    capacityPolicy = new CapacityVolumeChoosingPolicy();
    roundRobinPolicy = new RoundRobinVolumeChoosingPolicy();

    // Create NUM_VOLUMES volumes with different available space
    for (int i = 0; i < NUM_VOLUMES; i++) {
      String volumePath = baseDir + "/disk" + i;
      // Each volume has 1000 bytes capacity and different available space
      SpaceUsageSource source = MockSpaceUsageSource.fixed(1000000000, 1000000000 - i * 50);
      SpaceUsageCheckFactory factory = MockSpaceUsageCheckFactory.of(
          source, Duration.ZERO, SpaceUsagePersistence.None.INSTANCE);
      HddsVolume volume = new HddsVolume.Builder(volumePath)
          .conf(CONF)
          .usageCheckFactory(factory)
          .build();
      volumes.add(volume);
    }
  }

  // @AfterEach
  public void cleanUp() {
    volumes.forEach(HddsVolume::shutdown);
  }

  @Test
  public void testConcurrentVolumeChoosing() throws Exception {
    // Test both policies
    for (int i = 0; i < 5; i++) {
      setup();
      testPolicyPerformance("CapacityVolumeChoosingPolicy", capacityPolicy);
      cleanUp();

      setup();
      testPolicyPerformance("RoundRobinVolumeChoosingPolicy", roundRobinPolicy);
      cleanUp();
    }
  }

  private void testPolicyPerformance(String policyName, VolumeChoosingPolicy policy) throws Exception {
    CountDownLatch latch = new CountDownLatch(NUM_THREADS);
    AtomicInteger successCount = new AtomicInteger(0);
    AtomicInteger failureCount = new AtomicInteger(0);
    AtomicLong totalTimeNanos = new AtomicLong(0);

    // Submit NUM_THREADS tasks, each doing NUM_ITERATIONS volume choices
    for (int i = 0; i < NUM_THREADS; i++) {
      executor.submit(() -> {
        try {
          long threadStartTime = System.nanoTime();
          int success = 0;
          int failures = 0;
          for (int j = 0; j < NUM_ITERATIONS; j++) {
            try {
              HddsVolume volume = policy.chooseVolume(volumes, CONTAINER_SIZE);
              if (volume != null) {
                success++;
              }
            } catch (Exception e) {
              failures++;
            }
          }
          long threadEndTime = System.nanoTime();
          totalTimeNanos.addAndGet((threadEndTime - threadStartTime));
          successCount.addAndGet(success);
          failureCount.addAndGet(failures);
        } finally {
          latch.countDown();
        }
      });
    }

    // Wait for all tasks to complete
    assertTrue(latch.await(5, TimeUnit.MINUTES), "Test timed out");

    long totalOperations = (long) NUM_THREADS * NUM_ITERATIONS;
    double avgTimePerOperation = (double) totalTimeNanos.get() / totalOperations;
    double operationsPerSecond = totalOperations / (totalTimeNanos.get() / 1_000_000_000.0);

    System.out.println("Performance results for " + policyName);
    System.out.println("Total operations: " + totalOperations);
    System.out.println("Successful operations: " + successCount.get());
    System.out.println("Failed operations: " + failureCount.get());
    System.out.println("Total time: " + totalTimeNanos.get() / 1_000_000 + " ms");
    System.out.println("Average time per operation: " + avgTimePerOperation + " ns");
    System.out.println("Operations per second: " + operationsPerSecond);
    System.out.println("");
  }
}
```

peterxcli committed:
Use threadlocal random for Capacity and Remove AtomicInteger as CapacityVolumeChoosingPolicy is thread-safe now

peterxcli commented:
> Agree. We can do this in a new JIRA. Generally, the tar zip file of container will have smaller size than it's real size, so reserve a 2* max container size is conservative enough in most cases. But there are container over-allocated case, where container size can be double or triple the max container size as @siddhantsangwan saw in user's environment, have a accurate container size can handle this case welly.

Created one. https://issues.apache.org/jira/browse/HDDS-12998

peterxcli commented:
> i'm curious why capacity-based is several times slower than round-robin. But regardless, 12 thousand CreateContainer per second is more than enough.

It seems that these two lines are the cause:

[https://github.com/apache/ozone/blob/f52069b7aa757d39e4819bae7d06b12955660576/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java#L58-L60](https://github.com/apache/ozone/blob/f52069b7aa757d39e4819bae7d06b12955660576/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java#L58-L60)
[https://github.com/apache/ozone/blob/f52069b7aa757d39e4819bae7d06b12955660576/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java#L82-L83](https://github.com/apache/ozone/blob/f52069b7aa757d39e4819bae7d06b12955660576/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java#L82-L83)

Round-robin doesn't iterate or use randomness — in an ideal case, each call should just run `AvailableSpaceFilter` test and return `volumeSet[i++]`.

peterxcli commented:
Final benchmark result(NUM_VOLUMES=10, NUM_THREADS=100, NUM_ITERATIONS = 100000):
Here’s the data in **Markdown table** format for both *Average Latency (ns)* and *Operations Per Second (ops)*:

### Table

- Average Latency (ns):

|                    | Run1   | Run2   | Run3   | Run4   | Run5   |
|--------------------|--------|--------|--------|--------|--------|
| Capacity - sync    | 136095 | 138843 | 139845 | 138826 | 138220 |
| RoundRobin - sync  | 11754  | 11572  | 12911  | 11706  | 12895  |
| Capacity           | 85952  | 85972  | 86069  | 82412  | 80370  |
| RoundRobin         | 18425  | 17051  | 16194  | 16787  | 16410  |

- Operations Per Second (ops):

|                    | Run1  | Run2  | Run3  | Run4  | Run5  |
|--------------------|-------|-------|-------|-------|-------|
| Capacity - sync    | 7347  | 7202  | 7150  | 7203  | 7234  |
| RoundRobin - sync  | 85074 | 86414 | 77449 | 85422 | 77548 |
| Capacity           | 11634 | 11631 | 11618 | 12134 | 12442 |
| RoundRobin         | 54273 | 58647 | 61750 | 59566 | 60938 |

### Charts

<img width="1074" alt="image" src="https://github.com/user-attachments/assets/220739e4-8e1f-4eda-a99a-01f32b9b646e" />

<img width="1073" alt="image" src="https://github.com/user-attachments/assets/cd4bb754-5eac-4093-b9d8-d5b2182c4960" />

cc @ChenSammi

ChenSammi commented:
@peterxcli , the data looks same as the one I last shared.

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java:
```diff
@@ -214,21 +213,14 @@ public synchronized void setState(ContainerDataProto.State state) {
-      Preconditions.checkState(getMaxSize() > 0);
-      commitSpace();
-    }
+  public boolean isCommittedSpace() {
```
this method is used only in testcases, we can define @VisibleForTesting

peterxcli replied:
I'm not sure we really need @VisibleForTesting here, as it's just a getter, it won't affect anything if used in prod code.

sumitagrawl reviewed (commented):
@peterxcli Thanks for working over this, having few minor comments

peterxcli committed:
Release older and commit newer if container got conflict and swapped

ChenSammi commented:
> @ChenSammi I was asking gpt to generate the table and result for me, and I give the previous table image you shared as format example, and I didn't notice it use the data in it directly😥. Sorry I didn't check it.
> 
> BTW, I updated the test result, and I have no idea why synced choosing policies are performing better than no sync. Maybe [f52069b](https://github.com/apache/ozone/commit/f52069b7aa757d39e4819bae7d06b12955660576) really helped.

Yea, the optimization effect is really good.  Here is the data I collected for 100000000 operations. 

![Screenshot 2025-05-12 at 14 18 43](https://github.com/user-attachments/assets/1434f432-69ba-4fd2-a0ee-144d54083350)
![Screenshot 2025-05-12 at 14 18 33](https://github.com/user-attachments/assets/c80cffab-266d-46f1-89d6-c7f20cb79af3)

ChenSammi reviewed (approved):
The last patch LGTM. Thanks @peterxcli .  @siddhantsangwan , @sumitagrawl , would you like to take another look?

peterxcli commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerReader.java:
```diff
@@ -202,50 +202,52 @@ public void verifyAndFixupContainerData(ContainerData containerData)
+          throw e;
+        }
+        if (shouldDelete) {
+          resolveDuplicate((KeyValueContainer) containerSet.getContainer(
```
done.

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainer.java:
```diff
@@ -173,6 +173,8 @@ public void create(VolumeSet volumeSet, VolumeChoosingPolicy
           // volume to deduce the db file.
           containerData.setVolume(containerVolume);
+          // commit space has been reserved by volumeChoosingPolicy
+          containerData.setCommittedSpace(true);
```
@ChenSammi This part is pending, need move this logic at line 162 just after chooseVolume, so that in future if any code is added after that which may throw exception, it does not induce issue of not releasing commitedSpace in exception handler, as release depends on this flag to be "true"

https://github.com/apache/ozone/pull/8360/files#diff-c1cf18e174dcc55efd5f1177d6ea8e916353d2a59681388b225e97c512d9b0baL161:~:text=containerVolume-,%3D,-volumeChoosingPolicy.chooseVolume

sumitagrawl commented:
@ChenSammi One comment is pending to be fixed, after that is ok to be merged.

peterxcli committed:
Set commit space to true right after volume is choosen in KeyValueContainer

peterxcli committed:
Merge remote-tracking branch 'upstream/master' into hdds12810-make-volume-space-check-and-reservation-as-an-atomic-operation-in-volume-choosing-policy

peterxcli commented:
> @ChenSammi This part is pending, need move this logic at line 162 just after chooseVolume, so that in future if any code is added after that which may throw exception, it does not induce issue of not releasing commitedSpace in exception handler, as release depends on this flag to be "true"
> 
> https://github.com/apache/ozone/pull/8360/files#diff-c1cf18e174dcc55efd5f1177d6ea8e916353d2a59681388b225e97c512d9b0baL161:~:text=containerVolume-,%3D,-volumeChoosingPolicy.chooseVolume

Thanks for catching this! Done.

ChenSammi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainer.java:
```diff
@@ -159,6 +159,8 @@ public void create(VolumeSet volumeSet, VolumeChoosingPolicy
         try {
           containerVolume = volumeChoosingPolicy.chooseVolume(volumes, maxSize);
+          // commit bytes have been reserved in volumeChoosingPolicy#chooseVolume
+          containerData.setCommittedSpace(true);
```
@peterxcli , need move L174 & L177 before setCommittedSpace, otherwise, releaseCommitSpace will throw NPE for getVolume().

peterxcli committed:
Set volume first, too, to prevent releaseSpace throw NPE when getVolume

sumitagrawl reviewed (approved):
LGTM

peterxcli commented:
Thanks @ChenSammi for reviewing and optimizing and running the benchmark! And thanks @siddhantsangwan, @sumitagrawl for the review!
